components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServiceURL.java

```diff
@@ -0,0 +1,174 @@
+
+        return proxyContext;
+    }
+
+    public String getWebContext() {
+
+        return webContext;
+    }
+
+    public String getUrlContext() {
+
+        return urlContext;
+    }
+
+    public Map<String, String> getParameters() {
+
+        return parameters;
+    }
+
+    public Map<String, String> getFragments() {
+
+        return fragments;
+    }
+
+    private void appendContextToUri(StringBuilder serverUrl, String contextPath) {
+
+        if (StringUtils.isNotBlank(contextPath)) {
+            if (contextPath.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(contextPath.trim());
+            } else {
+                serverUrl.append(contextPath.trim());
+            }
+        }
+    }
+
+    private void appendParamsToUri(StringBuilder serverUrl, Map<String, String> parameters) {
+
+        if (MapUtils.isNotEmpty(parameters)) {
+            for (Map.Entry<String, String> entry : parameters.entrySet()) {
+                serverUrl.append("?").append(entry.getKey()).append("=").append(entry.getValue());
+            }
+        }
+    }
+
+    /**
+     * Concatenate the protocol, host name, port, proxy context path, web context root and url context to return the
+     * absolute URL.
+     *
+     * @return The absolute URL from the {@link ServiceURL} instance.
+     */
+    public String getAbsoluteUrl() {
+
+        StringBuilder absoluteUrl = new StringBuilder();
+        absoluteUrl.append(getProtocol()).append("://");
+
+        String hostName = getHostName();
+        if (hostName.endsWith("/")) {
+            hostName = hostName.substring(0, hostName.length() - 1);
+        }
+        absoluteUrl.append(hostName.toLowerCase());
+
+        int port = getPort();
+        // If it's well known HTTPS port, skip adding port.
+        if (port != IdentityCoreConstants.DEFAULT_HTTPS_PORT) {
+            absoluteUrl.append(":").append(port);
+        }
+
+        String proxyContextPath = getProxyContext();
+        appendContextToUri(absoluteUrl, proxyContextPath);
+
+        String webContextRoot = getWebContext();
+        appendContextToUri(absoluteUrl, webContextRoot);
+
+        String urlContext = getUrlContext();
+        appendContextToUri(absoluteUrl, urlContext);
+
+        Map<String, String> parameters = getParameters();
+        appendParamsToUri(absoluteUrl, parameters);
+
+        if (absoluteUrl.toString().endsWith("/")) {
+            absoluteUrl.setLength(absoluteUrl.length() - 1);
+        }
+        return absoluteUrl.toString();
+    }
+
+    /**
+     * This method is called to get the relative url from the url context.
+     *
+     * @return The relative URL from the {@link ServiceURL} instance.
+     */
+    public String getRelativeUrl() {
+
+        StringBuilder relativeUrl = new StringBuilder();
+        String urlContext = getUrlContext();
+        appendContextToUri(relativeUrl, urlContext);
+
+        Map<String, String> parameters = getParameters();
```

Comment:
Need to handle URL fragements

Reply:
Should we add fragments to relative URL as well?